[SUPPORTESC-280] Zapier constraints section in Platform documentation (#271)
zapier/visual-builder: docs/_docs/constraints.md

**Example:** The [Salesforce](https://zapier.com/apps/salesforce/integrations) integration has an **_Updated Field on Record_** trigger. It can fire on older records that a user might not consider appropriate for their running Zap. To help mitigate this, the integration uses a distinct filtered request to download up to 104,000 records to the deduplication table during Zap startup.

# Zap step execution

> Once a Zap is enabled, then a different set of constraints and best practices come into play. This is true of the Zapier run-time, but also usually true of your integration and its API requests. We’ll touch on both.

Author: I like this!

## Timeouts (triggers)

**Constraint:** Each time a Zap executes the trigger's `perform` (polling) method must finish processing in 30 seconds. Polling triggers run on an interval based on a user's Zapier plan, REST Hook triggers run on an inbound POST to their subscription URL.

**What errors a user will see if constraint hit:**

- User will receive an email with an error message, usually with _"Trigger Partner Failure"_ in the message text. An example of the email sent to when the trigger errors due to a timeout:

![](https://cdn.zappy.app/17cc8fb9313f8b0920a7ccd6dd95e9b0.png)

**Best practice:**

> - For polling triggers, reducing the number of records per poll (much like the sample example earlier)

Author: I can find the direct link here (at least for platform), so I'll add it once I re-render the file.

- For both types of triggers, optimizing `perform` scripting you have for manipulating the payload
- Using [console logging](https://platform.zapier.com/cli_docs/docs#console-logging) can help you determine where the issue might lie. One thing to look out for here, too much console logging can also cause timeouts due to round trip HTTP calls for each log message.
- If you have multiple requests per record that are causing timeouts, another option is to use the Zapier platform dehydration service, as explained [here](https://platform.zapier.com/cli_docs/docs#dehydration). Instead of making the request directly, a dehydration pointer is created, and this pointer will only be resolved if the Zap needs a hydrated property in a later step.

> - Dehydration provides one other advantage - if there is an error in trying to hydrate the request, the error will be exposed in Zap History and the Zap can be replayed from there.

Author: the thought here is if you do dehydration, a secondary benefit is that you can replay if there is an error. But, it might be better in that section.

## Timeouts (actions)

**Constraint:** Sometimes an API request will never be finished within 30 seconds - the rendering of a file conversion is a good example of this.

**What errors a user will see if constraint hit:**

- An error in the Zap history of their Zap due to the request timing out

**Best practice:** To help with this kind of problem, the Zapier platform provides a webhook-based callback service. This allows your action to be performed asynchronously, and when finished, POST to the callback URL. The Zap will be put in waiting mode until this POST. More on utilizing this method [here](https://platform.zapier.com/cli_docs/docs#zgeneratecallbackurl).

**What a user will see:**

- Zap will have Waiting/Delayed status in Zap history until the POST is received.

## Payload size (actions/searches)

**Constraint:** When a user clicks **Test and Review** or **Retest and Review** in the Zap editor your response payload must be less than 6MB.

**What error messages a user could see if constraint hit:**

- _“Scripting payload too large ('n' bytes but max is 6291456bytes).”_

> **Best practice:** If your API endpoints supports this, one option is to provide input filter fields in the action/search so a user can decide what record data they want to return. If that is not available, you might look at having multiple endpoints for this record data. Your integration could then provide multiple action/searches for the user so they can get a full record.

Reviewer: Similar comment to the others about "input filter fields"

## Reducing requests to your API

**Constraint:** Each time a Zap runs and your integration is invoked can mean multiple requests to your API endpoints, depending on what your trigger/action seeks to accomplish. This can mean hundreds of requests/hour based on a Zapier customer plan, or thousands based on a Zap with an active REST Hook Trigger.

**Best practice:** One way to reduce that API load is via the Zapier platform dehydration feature mentioned earlier. By putting these secondary requests behind a dehydration pointer, Zapier will only make this request once, although it might see these same records again and again based on the Zap’s polling cycle.

More on dehydration again [here](https://platform.zapier.com/cli_docs/docs#dehydration).

## Reducing file requests to your API

**Constraint:** Each time a Zap step requests a file from your API, it will be accessed and downloaded.

> **Best practice:** Much like data dehydration, you can implement a file request using [dehydration](https://platform.zapier.com/cli_docs/docs#zgeneratecallbackurl#file-dehydration), so again the file will only be accessed and downloaded when a later Zap step asks for it.

Reviewer: ```suggestion
**Best practice:** Much like data dehydration, you can implement a file request using [dehydration](https://platform.zapier.com/cli_docs/docs#zgeneratecallbackurl#file-dehydration), so the file will only be accessed and downloaded when a later Zap step asks for it.
```

To make this even more efficient, you can [stash the file](https://platform.zapier.com/cli_docs/docs#stashing-files) at Zapier. Rather than provide the file to the requesting step, Zapier will stash the file (under a dehydrated URL), so that only one request will ever be made for the file from your API.

# Throttling

There are a number of throttles that a Zapier user could encounter when using your integration, including those set by your API.

## Throttling (your API)

**Constraint:** Your API has request limits

> **What errors a user will see if constraint hit:**

Reviewer: ```suggestion
**Errors a user will see if constraint is hit:**
```

- If a trigger, user will receive an email with an error message about the trigger error
- If an action, user will see an error in Zap history

**Best practice:** You can add a specific 429 `Retry-After` header to your response, or add a timed delay in your error response. Instead of a user’s Zap erroring and halting, the request will be retried at the specified time. More on the retry [here](https://platform.zapier.com/cli_docs/docs#handling-throttled-requests). The user will see this message in Zap History instead of an error while the limit is still in place:
![](https://cdn.zappy.app/0933736266259b11771a0eba0aff23ce.png)

## Webhook throttles (Zapier)

**Constraint:** Zapier’s current webhook limits are [here](https://zapier.com/help/troubleshoot/behavior/rate-limits-and-throttling-in-zapier#step-4). We will issue a 429 response when your integration exceeds these throttle limits.

> **What errors a user will see:**

Reviewer: ```suggestion
**Errors a user will see if constraint is hit:**
```

- If a trigger, user will receive an email with an error message about the trigger error
- If an action, user will see error in Zap history

**Best practice:** You should support a retry/back-off schedule to make sure the data is eventually received.

## Polling trigger throttle (Zapier)

> **Constraint:** There is a default limit of 100 returned items/poll.

Reviewer: ```suggestion
**Constraint:** There is a default limit of 100 new items/poll after deduplication.
```

Reviewer: I'm thinking we might actually want to link directly to the held runs docs here, as those writing may not yet have come across it.

> **What errors a user will see if constraint hit:**

Reviewer: ```suggestion
**Errors a user will see if constraint is hit:**
```

- The user will receive an email about held Zap runs, as well as a banner with the same information in their Zap history.

> **Best practice:** If your trigger will be returning > 100 new records consistently, we’d encourage you to look at these two options:

Reviewer: ```suggestion
**Best practices:** If your trigger will be returning > 100 new records consistently, we’d encourage you to look at these two options:
```

- Converting your trigger to be REST Hook based. Webhook limits are much expanded (up to 10,000 requests in a 5 minute period).

> - Add support for Transfer. This is a new Zapier service that supports paging so users can create a Zap tp select and batch transfer up to 25,000 records at one time. More [here](https://platform.zapier.com/docs/transfer).

Reviewer: ```suggestion
- Add support for Transfer. This is a new Zapier service that supports paging so users can create a Zap to select and batch transfer up to 25,000 records at one time. More [here](https://platform.zapier.com/docs/transfer).
```
